src/classes/RD_ProcessControl.cls

```diff
@@ -38,11 +38,28 @@ public with sharing class RD_ProcessControl {
     /*******************************************************************************************************
     * @description flag to prevent recursion into the RD trigger handler/logic
     */ 
-    public static boolean hasRun = false;
+    private static Integer recursionFlag = 0;
```

Comment:
so we've now encountered the recursionFlag, setRecursionFlag(), checkRecursionState() in many of our process control classes.  why not create a base process control class that these all inherit from that provides this implementation so it is truly shared.  then individual process control classes can add any additional logic or additional flags as needed.  but this way the common way is shared.